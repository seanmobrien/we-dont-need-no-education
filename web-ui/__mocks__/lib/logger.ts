/**
 * @file This module provides a mock implementation of a logger for testing purposes.
 *
 * The `logger` function returns an object with various logging methods (`warn`, `error`, `info`, `debug`, `silly`, `verbose`, `log`),
 * each of which is a Jest mock function. These mock functions log their calls to the console with a specific format,
 * indicating the method name and the arguments passed to it.
 *
 * This mock logger can be used in unit tests to verify that logging methods are called with the expected arguments
 * without actually performing any logging operations.
 */
const makeMockImplementation = (name: string) => {
  return (...args: unknown[]) =>
    console.log(`logger::${name} called with `, args);
};
export const logger = () => ({
  warn: jest.fn(makeMockImplementation('warn')),
  error: jest.fn(makeMockImplementation('error')),
  info: jest.fn(makeMockImplementation('info')),
  debug: jest.fn(makeMockImplementation('debug')),
  silly: jest.fn(makeMockImplementation('silly')),
  verbose: jest.fn(makeMockImplementation('verbose')),
  log: jest.fn(makeMockImplementation('log')),
});
export const log = jest.fn((cb: (l: ReturnType<typeof logger>) => void) =>
  cb(logger())
);
