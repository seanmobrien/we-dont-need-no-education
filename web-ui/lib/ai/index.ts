export * from './core';
export type * from './types';
export * from './aiModelFactory';
export * from './guards';
export * from './tools';
