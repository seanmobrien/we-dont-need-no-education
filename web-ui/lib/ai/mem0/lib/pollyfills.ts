import { env } from '/lib/site-util/env';

/**
 * Overrides default behavior of @ai-sdk/provider-utils loadApiKey to account for the fact that, well, we're not using an API key.
 * @param param0
 * @returns
 */
export const loadApiKey = ({
  apiKey,
  environmentVariableName,
  apiKeyParameterName,
}: {
  apiKey: string | undefined;
  environmentVariableName: string;
  apiKeyParameterName?: string;
  description?: string;
}): string =>
  apiKey ??
  process.env[environmentVariableName] ??
  (apiKeyParameterName ? process.env[apiKeyParameterName] : undefined) ??
  '';

/**
 * Helper function to construct a mem0 api url.
 * @param relativePath - The relative path to append to the base URL.
 * @returns The full API URL.
 */
export const getMem0ApiUrl = (relativePath: string): string => {
  const baseUrl = env('MEM0_API_HOST');
  return new URL(relativePath, baseUrl).toString();
};
