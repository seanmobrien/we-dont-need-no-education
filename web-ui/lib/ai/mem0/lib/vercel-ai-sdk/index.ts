export * from './src/mem0-facade';
export type { Mem0Provider, Mem0ProviderSettings } from './src/mem0-provider';
export { createMem0, mem0 } from './src/mem0-provider';
export type {
  Mem0ConfigSettings,
  Mem0ChatConfig,
  Mem0ChatSettings,
} from './src/mem0-types';
export {
  addMemories,
  retrieveMemories,
  searchMemories,
  getMemories,
} from './src/mem0-utils';
