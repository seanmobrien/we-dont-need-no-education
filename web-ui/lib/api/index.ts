export { RepositoryCrudController } from './repository-crud-controller';
export type * from './_types';
export * from './_baseObjectRepository';
export * from './thread/database';
export * from './email';
export type * from './email';
