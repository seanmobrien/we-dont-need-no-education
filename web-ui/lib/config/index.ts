export { withIgnorePacks } from './ignore-unsupported-packs-plugin';
export { withStripRscPrefixPlugin } from './strip-rsc-prefix-plugin';
export { withPublicEnv } from './public-env';
export { withBundleAnalyzer } from './bundle-analyzers';