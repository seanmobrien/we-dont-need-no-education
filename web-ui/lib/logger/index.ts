import winston from 'winston';

const ServerLoggerId = 'nextjs-server';
const ClientLoggerId = 'nextjs-client';

const getServerSideLogger = () => {
  let ret = winston.loggers.get(ServerLoggerId);
  if (ret) {
    return ret;
  }
  ret = winston.loggers.add(ServerLoggerId, {
    level: process.env.LOG_LEVEL_SERVER || 'error',
    format: winston.format.combine(
      winston.format.errors({ stack: true }),
      winston.format.timestamp(),
      winston.format.json()
    ),
    defaultMeta: { service: ServerLoggerId },
    transports: [
      new winston.transports.Console({
        format: winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.timestamp(),
          winston.format.colorize(),
          winston.format.simple()
        ),
      }),
    ],
    exceptionHandlers: [
      new winston.transports.File({ filename: 'exception.log' }),
    ],
    rejectionHandlers: [
      new winston.transports.File({ filename: 'rejections.log' }),
    ],
  });
  ret.verbose('Server logger created for process %s', process.pid);
  return ret;
};
const getClientSideLogger = () => {
  let ret = winston.loggers.get(ClientLoggerId);
  if (ret) {
    return ret;
  }
  ret = winston.loggers.add(ClientLoggerId, {
    level: process.env.LOG_LEVEL_CLIENT || 'error',
    format: winston.format.combine(
      winston.format.errors({ stack: true }),
      winston.format.timestamp(),
      winston.format.json()
    ),
    defaultMeta: { service: ClientLoggerId },
    transports: [
      new winston.transports.Console({
        format: winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.timestamp(),
          winston.format.colorize(),
          winston.format.simple()
        ),
      }),
    ],
    exceptionHandlers: [
      new winston.transports.File({ filename: 'exception.log' }),
    ],
    rejectionHandlers: [
      new winston.transports.File({ filename: 'rejections.log' }),
    ],
  });
  ret.verbose('Client logger created for process %s', process.pid);
  return ret;
};

export const logger = (id?: string) => {
  if (id) {
    return winston.loggers.get(id);
  }
  return typeof window === 'undefined'
    ? getServerSideLogger()
    : getClientSideLogger();
};
