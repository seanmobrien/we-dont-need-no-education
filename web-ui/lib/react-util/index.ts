/**
 * @module react-util
 *
 * This module serves as a central point to access various utility functions used
 * across the React application.
 */
export * from './_utility-methods';
export * from './errors';
export * from './hooks';
