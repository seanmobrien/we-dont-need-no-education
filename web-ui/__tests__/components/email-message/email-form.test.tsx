/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import EmailForm from 'components/email-message/email-form';
import { logger } from 'lib/logger';
import { generateUniqueId } from 'lib/react-util';

jest.mock('lib/logger');
jest.mock('lib/react-util');

describe('EmailForm', () => {
  const mockFetch = jest.fn();

  beforeAll(() => {
    global.fetch = mockFetch;
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the form with initial values', () => {
    const { asFragment } = render(<EmailForm />);
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByLabelText(/Sender ID/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Subject/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Email Contents/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Sent Timestamp/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Thread ID/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Parent Email ID/i)).toBeInTheDocument();
  });

  it('should display a message when email is created successfully', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ message: 'Email created successfully!' }),
    });

    render(<EmailForm />);

    fireEvent.change(screen.getByLabelText(/Sender ID/i), {
      target: { value: '1' },
    });
    fireEvent.change(screen.getByLabelText(/Subject/i), {
      target: { value: 'Test Subject' },
    });
    fireEvent.change(screen.getByLabelText(/Email Contents/i), {
      target: { value: 'Test Body' },
    });
    fireEvent.change(screen.getByLabelText(/Sent Timestamp/i), {
      target: { value: '2023-01-01T00:00' },
    });

    fireEvent.click(screen.getByRole('button', { name: /Create Email/i }));

    await waitFor(() => {
      expect(
        screen.getByText(/Email created successfully!/i)
      ).toBeInTheDocument();
    });
  });

  it('should display an error message when email creation fails', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: false,
      json: async () => ({ error: 'Something went wrong.' }),
    });

    render(<EmailForm />);

    fireEvent.change(screen.getByLabelText(/Sender ID/i), {
      target: { value: '1' },
    });
    fireEvent.change(screen.getByLabelText(/Subject/i), {
      target: { value: 'Test Subject' },
    });
    fireEvent.change(screen.getByLabelText(/Email Contents/i), {
      target: { value: 'Test Body' },
    });
    fireEvent.change(screen.getByLabelText(/Sent Timestamp/i), {
      target: { value: '2023-01-01T00:00' },
    });

    fireEvent.click(screen.getByRole('button', { name: /Create Email/i }));

    await waitFor(() => {
      expect(screen.getByText(/Something went wrong./i)).toBeInTheDocument();
    });
  });

  it('should display a network error message when fetch fails', async () => {
    mockFetch.mockRejectedValueOnce(new Error('Network error'));

    render(<EmailForm />);

    fireEvent.change(screen.getByLabelText(/Sender ID/i), {
      target: { value: '1' },
    });
    fireEvent.change(screen.getByLabelText(/Subject/i), {
      target: { value: 'Test Subject' },
    });
    fireEvent.change(screen.getByLabelText(/Email Contents/i), {
      target: { value: 'Test Body' },
    });
    fireEvent.change(screen.getByLabelText(/Sent Timestamp/i), {
      target: { value: '2023-01-01T00:00' },
    });

    fireEvent.click(screen.getByRole('button', { name: /Create Email/i }));

    await waitFor(() => {
      expect(
        screen.getByText(/Network error. Please try again./i)
      ).toBeInTheDocument();
    });
  });

  it('should fetch and populate email details when editing', async () => {
    const mockEmailData = {
      sender_id: 1,
      subject: 'Test Subject',
      body: 'Test Body',
      sent_timestamp: '2023-01-01T00:00:00Z',
      thread_id: 2,
      parent_email_id: 3,
    };

    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockEmailData,
    });

    render(<EmailForm emailId={1} />);

    await waitFor(() => {
      expect(document.getElementById('senderId-unique-id-1')).toHaveValue(1);
      expect(document.getElementById('subject-unique-id-1')).toHaveValue(
        'Test Subject'
      );
      expect(document.getElementById('emailContents-unique-id-1')).toHaveValue(
        'Test Body'
      );
      expect(
        document.getElementById('sentTimestamp-unique-id-1')
      ).toHaveAttribute('value', '2023-01-01T00:00:00Z');
      expect(document.getElementById('threadId-unique-id-1')).toHaveValue(2);
      expect(document.getElementById('parentEmailId-unique-id-1')).toHaveValue(
        3
      );
    });
  }, 5000);

  it('should display an error message when fetching email details fails', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: false,
      json: async () => ({ error: 'Error fetching email details.' }),
    });

    render(<EmailForm emailId={1} />);

    await waitFor(() => {
      expect(
        screen.getByText(/Error fetching email details./i)
      ).toBeInTheDocument();
    });
  }, 3000);
});
