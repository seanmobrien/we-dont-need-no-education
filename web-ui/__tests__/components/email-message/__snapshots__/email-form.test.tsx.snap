// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailForm should render the form with initial values 1`] = `
<DocumentFragment>
  <div
    class="max-w-lg mx-auto p-6 bg-white rounded-lg shadow-md"
  >
    <h2
      class="mb-4 text-xl font-semibold"
    >
      Create Email
    </h2>
    <form
      class="space-y-4"
    >
      <div>
        <label
          class="block font-medium mb-1"
          for="senderId-unique-id-1"
          id="senderIdLabel-unique-id-1"
        >
          Sender ID
        </label>
        <input
          aria-labelledby="senderIdLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="senderId-unique-id-1"
          required=""
          type="number"
          value=""
        />
      </div>
      <div>
        <label
          class="block font-medium mb-1"
          for="subject-unique-id-1"
          id="subjectLabel-unique-id-1"
        >
          Subject
        </label>
        <input
          aria-labelledby="subjectLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="subject-unique-id-1"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          class="block font-medium mb-1"
          for="emailContents-unique-id-1"
          id="emailContentsLabel-unique-id-1"
        >
          Email Contents
        </label>
        <textarea
          aria-labelledby="emailContentsLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="emailContents-unique-id-1"
          required=""
        />
      </div>
      <div>
        <label
          class="block font-medium mb-1"
          for="sentTimestamp-unique-id-1"
          id="sentTimestampLabel-unique-id-1"
        >
          Sent Timestamp
        </label>
        <input
          aria-labelledby="sentTimestampLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="sentTimestamp-unique-id-1"
          required=""
          type="datetime-local"
          value=""
        />
      </div>
      <div>
        <label
          class="block font-medium mb-1"
          for="threadId-unique-id-1"
          id="threadIdLabel-unique-id-1"
        >
          Thread ID (optional)
        </label>
        <input
          aria-labelledby="threadIdLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="threadId-unique-id-1"
          type="number"
          value=""
        />
      </div>
      <div>
        <label
          class="block font-medium mb-1"
          for="parentEmailId-unique-id-1"
          id="parentEmailIdLabel-unique-id-1"
        >
          Parent Email ID (optional)
        </label>
        <input
          aria-labelledby="parentEmailIdLabel-unique-id-1"
          class="w-full border rounded p-2 focus:outline-none focus:ring focus:ring-blue-300"
          id="parentEmailId-unique-id-1"
          type="number"
          value=""
        />
      </div>
      <button
        aria-roledescription="Submit Form"
        class="w-full p-2 rounded text-white hover:opacity-80 transition bg-blue-500 hover:bg-blue-600 disabled:bg-gray-400"
        type="submit"
      >
        Create Email
      </button>
    </form>
  </div>
</DocumentFragment>
`;
